### ISSUE TITLE
Inconsistent behavior for in-place operations on coalesced sparse tensors

### ISSUE THREAD

**Issue by jeromerony** [pytorch/pytorch#109462]
### 🐛 Describe the bug

When working with coalesced sparse COO tensors, calling `.div_` keeps the tensor coalesced, but `.mul_` does not. However, calling `.mul` returns a coalesced tensor.

```python
>>> i = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v = torch.tensor([3, 4, 5], dtype=torch.float32)
>>> t = torch.sparse_coo_tensor(i, v, [2, 4]).coalesce()
>>> t.mul(2).is_coalesced()
True
>>> t.div(2).is_coalesced()
True
>>> t.div_(2).is_coalesced()
True
>>> t.mul_(2).is_coalesced()
False
```

### Versions

Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Amazon Linux 2 (x86_64)
GCC version: (GCC) 7.3.1 20180712 (Red Hat 7.3.1-17)
Clang version: Could not collect
CMake version: version 2.8.12.2
Libc version: glibc-2.26

Python version: 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:17) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.10.178-162.673.amzn2.x86_64-x86_64-with-glibc2.26
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A10G
Nvidia driver version: 525.85.12
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              16
On-line CPU(s) list: 0-15
Thread(s) per core:  2
Core(s) per socket:  8
Socket(s):           1
NUMA node(s):        1
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               49
Model name:          AMD EPYC 7R32
Stepping:            0
CPU MHz:             3028.998
BogoMIPS:            5599.99
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-15
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch topoext ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr rdpru wbnoinvd arat npt nrip_save rdpid

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] robust-loss-pytorch==0.0.2
[pip3] torch==2.0.1
[pip3] torch-dct==0.1.6
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] blas                      2.116                       mkl    conda-forge
[conda] blas-devel                3.9.0            16_linux64_mkl    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            16_linux64_mkl    conda-forge
[conda] mkl                       2022.1.0           h84fe81f_915    https://aws-ml-conda-pre-prod-ec2.s3.us-west-2.amazonaws.com
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numexpr                   2.8.3           mkl_py311h0b1cd97_1    conda-forge
[conda] numpy                     1.24.4          py311h64a7726_0    conda-forge
[conda] pytorch                   2.0.1           py3.11_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_3    https://aws-ml-conda-pre-prod-ec2.s3.us-west-2.amazonaws.com
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] robust-loss-pytorch       0.0.2                    pypi_0    pypi
[conda] torch-dct                 0.1.6                    pypi_0    pypi
[conda] torchtriton               2.0.0                     py311    pytorch
[conda] torchvision               0.15.2              py311_cu118    pytorch

**Comment by nikitaved** [comment]
The problem is `add_`. Have a look at the logic:

https://github.com/pytorch/pytorch/blob/23938640706214921c0c0c29b1ed4855a0536cd8/aten/src/ATen/native/sparse/SparseTensorMath.cpp#L822-L832
`res` is not coalesced and we have:
```
In [2]: import torch

In [3]: >>> i = torch.tensor([[0, 1, 1], [2, 0, 2]])
   ...: >>> v = torch.tensor([3, 4, 5], dtype=torch.float32)
   ...: >>> t = torch.sparse_coo_tensor(i, v, [2, 4])

In [4]: t.is_coalesced()
Out[4]: False

In [5]: t.coalesce().add_(t).is_coalesced()
Out[5]: False
```

**Comment by nikitaved** [comment]
@jeromerony , you are right, sorry, hit wrong path :) But let's add `add_` to the repro as well?

**Comment by jeromerony** [comment]
I feel like the `add_` is a different issue, because it has to account for more cases: `add` can result in non coalesced tensors, but `add_` should not.
In the case of `mul_`,  the issue is the inconsistency with `mul`, `div` and `div_`.

**Comment by nikitaved** [comment]
I would say the general rule is that in-place ops on coalesced inputs should be coalesced. It could be we forgot to set the coalesce flag accordingly...

**Comment by jeromerony** [comment]
Agreed. Btw, it seems `add_` does result in a coalesced tensor if the other tensor is coalesced, even if they have different indices.

**Comment by nikitaved** [comment]
@jeromerony , you mean it produces false-coalesce inputs? IIRC, `add` just concatenates indices/values, so the result is not coalesced in general, unless this logic is handled with care, but I am not sure...

**Comment by jeromerony** [comment]
That does not appear to be the case:
```python
>>> i = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v = torch.tensor([3, 4, 5], dtype=torch.float32)
>>> t = torch.sparse_coo_tensor(i, v, [2, 4]).coalesce()
>>> i2 = torch.tensor([[0, 0, 1], [2, 0, 2]])
>>> v2 = torch.tensor([3, 4, 5], dtype=torch.float32)
>>> t2 = torch.sparse_coo_tensor(i2, v2, [2, 4]).coalesce()
>>> t.to_dense()
tensor([[0., 0., 3., 0.],
        [4., 0., 5., 0.]])
>>> t2.to_dense()
tensor([[4., 0., 3., 0.],
        [0., 0., 5., 0.]])
>>> t.add(t2).is_coalesced()
True
>>> t.add(t2).indices()
tensor([[0, 0, 1, 1],
        [0, 2, 0, 2]])
>>> t.add_(t2).is_coalesced()
True
>>> t.indices()
tensor([[0, 0, 1, 1],
        [0, 2, 0, 2]])
```
